src/plugins/MediaManager.js

```diff
@@ -0,0 +1,352 @@
+'use strict';
+
+const log = require('loglevel');
+log.setDefaultLevel('debug');
+
+module.exports = class MediaManager {
+    /**
+     * Constructor to this MediaManager class
+     * @param {DeviceRenderer} instance root instance
+     * @param {boolean} videoWithMicrophone flag indicating wether the video must always be played with the microphone.
+     *  This flag is used when we don't want to split audio&video and instead want the audio bundled with the video in the same stream.
+     *  Defaults to the microphone flag in the instance options
+     * @param {number} videoWidth Maximum video width. Defaults to 1280
+     * @param {number} videoHeight Maximum video height. Defaults to 720
+     */
+    constructor(instance, videoWithMicrophone = instance.options.microphone, videoWidth = 1280, videoHeight = 720) {
+        this.instance = instance;
+        this.instance.mediaManager = this;
+
+        if (!navigator.mediaDevices) {
+            log.error('MediaDevices API unsupported: camera and microphone won\'t be available.');
+            this.instance.mediaEventsEnabled = false;
+            return;
+        }
+        this.setupPermissions();
+        this.instance.mediaEventsEnabled = true;
+
+        this.videoStreaming = false;
+        this.audioStreaming = false;
+
+        this.videoWidth = videoWidth;
+        this.videoHeight = videoHeight;
+        this.videoWithMicrophone = videoWithMicrophone;
+    }
+
+    /**
+     * Tries to get the current microphone and camera permission status. If it can, binds listeners to their change
+     * @returns {boolean} True if it could find the permissions object
+     */
+    async setupPermissions() {
+        if (!navigator.permissions) {
+            return false;
```

Comment:
Shouldn't you set `this.instance.mediaEventsEnabled = false` here? In fact, this field seems totally unused

Reply:
`mediaEventsEnabled` is to be used by other classes above this manager (for example for disabling the UI), so that's why it is unused in this class.

It is set to `true` on line 26, as long as we're able to use the `mediaDevices` API. I don't want to set it to false if we can't use the `permissions` API because the feature still works, it simply won't be able to react dynamically to permission changes.

Reply:
Thanks for the explanation